pkg/operator/ceph/object/user/controller.go

```diff
@@ -0,0 +1,229 @@
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc: c.onAdd,
+		// UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching object store user resources in namespace %s", namespace)
+	watcher := opkit.NewWatcher(ObjectStoreUserResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CephV1beta1().RESTClient())
+	go watcher.Watch(&cephv1beta1.ObjectStoreUser{}, stopCh)
+
+	return nil
+}
+
+func (c *ObjectStoreUserController) onAdd(obj interface{}) {
+	user, err := getObjectStoreUserObject(obj)
+	if err != nil {
+		logger.Errorf("failed to get objectstoreuser object: %+v", err)
+		return
+	}
+
+	if err = c.createUser(c.context, user); err != nil {
+		logger.Errorf("failed to create object store user %s. %+v", user.Name, err)
+	}
+}
+
+// TODO: Need to think through how to do updates, leaving out for now.
+
+// func (c *ObjectStoreUserController) onUpdate(oldObj, newObj interface{}) {
+// 	// if the object store user spec is modified, update the object store user
+// 	oldStoreUser, err := getObjectStoreUserObject(oldObj)
+// 	if err != nil {
+// 		logger.Errorf("failed to get old objectstoreuser object: %+v", err)
+// 		return
+// 	}
+// 	newStoreUser, err := getObjectStoreUserObject(newObj)
+// 	if err != nil {
+// 		logger.Errorf("failed to get new objectstoreuser object: %+v", err)
+// 		return
+// 	}
+//
+// 	if !storeUserChanged(oldStoreUser.Spec, newStoreUser.Spec) {
+// 		logger.Debugf("object store user %s did not change", newStoreUser.Name)
+// 		return
+// 	}
+//
+// 	logger.Infof("applying object store user %s changes", newStoreUser.Name)
+// 	if err = UpdateStoreUser(c.context, *newStoreUser, c.rookImage, c.hostNetwork, c.storeUserOwners(newStoreUser)); err != nil {
+// 		logger.Errorf("failed to create (modify) object store user %s. %+v", newStoreUser.Name, err)
+// 	}
+// }
+
+func (c *ObjectStoreUserController) onDelete(obj interface{}) {
+	user, err := getObjectStoreUserObject(obj)
+	if err != nil {
+		logger.Errorf("failed to get objectstoreuser object: %+v", err)
+		return
+	}
+
+	if err = deleteUser(c.context, user); err != nil {
+		logger.Errorf("failed to delete object store user %s. %+v", user.Name, err)
+	}
+}
+
+func (c *ObjectStoreUserController) storeUserOwners(store *cephv1beta1.ObjectStoreUser) []metav1.OwnerReference {
+	// Only set the cluster crd as the owner of the object store user resources.
+	// If the object store user crd is deleted, the operator will explicitly remove the object store user resources.
+	// If the object store user crd still exists when the cluster crd is deleted, this will make sure the object store user
+	// resources are cleaned up.
+	return []metav1.OwnerReference{c.ownerRef}
+}
+
+func getObjectStoreUserObject(obj interface{}) (objectstoreuser *cephv1beta1.ObjectStoreUser, err error) {
+	var ok bool
+	objectstoreuser, ok = obj.(*cephv1beta1.ObjectStoreUser)
+	if ok {
```

Comment:
Have you seen a need to check if the cast was successful? I believe everywhere else we assume the cast to the CRD type is successful and we haven't ever seen an issue with that. This is one of the few cases where I ever consider skipping error checking so I'm not saying to skip error checking in general.

Reply:
I've not personally seen any need to check in my testing. I noticed that the [object controller](https://github.com/rook/rook/blob/7121418cef1fc36923f991e6a4949acd819dcf50/pkg/operator/ceph/object/controller.go#L218-L224) has the same logic, but perhaps that is only necessary because it could need to be upgraded. I can definitely remove this function if it is unnecessary.

Reply:
Ah, this pattern was more common than I remembered since we had to upgrade from the alpha types in our last release. It is fine to leave it since we may need it to upgrade to v1 types in the future.